distribution/tools/entitlement-agent/src/main/java/org/elasticsearch/entitlement/instrumentation/Instrumenter.java

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.entitlement.instrumentation;
+
+import org.elasticsearch.entitlement.checks.EntitlementChecks;
+import org.elasticsearch.entitlement.checks.EntitlementProvider;
+import org.objectweb.asm.AnnotationVisitor;
+import org.objectweb.asm.ClassReader;
+import org.objectweb.asm.ClassVisitor;
+import org.objectweb.asm.ClassWriter;
+import org.objectweb.asm.MethodVisitor;
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.Type;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.reflect.Method;
+import java.util.Map;
+
+import static org.objectweb.asm.ClassWriter.COMPUTE_FRAMES;
+import static org.objectweb.asm.ClassWriter.COMPUTE_MAXS;
+import static org.objectweb.asm.Opcodes.ACC_STATIC;
+import static org.objectweb.asm.Opcodes.GETSTATIC;
+import static org.objectweb.asm.Opcodes.INVOKEINTERFACE;
+import static org.objectweb.asm.Opcodes.INVOKESTATIC;
+import static org.objectweb.asm.Opcodes.INVOKEVIRTUAL;
+
+public class Instrumenter {
+    /**
+     * To avoid class name collisions during testing. Should be an empty string in production.
```

Comment:
Actually I removed the unit test that had this problem and forgot to put it back in. I should do that.

Here's the reason. I can add more of this info to the code if you want.

Developing the instrumentation code is tricky. Failures in the agent, or bad bytecodes, generally defeat IntelliJ and are hard to single-step; in fact, even the exception stack traces are often missing or incomplete, and it's just generally hard to tell what went wrong.

This led me to develop a unit test that tests the instrumentation logic sans agent. The trouble is, without the agent, you can't replace an existing class. This means the instrumentation needs to generate a new class, and I achieved that by allowing the test to specify a name suffix.

Reply:
Is the instrumenter meant to handle a single class, or one instrumenter for all classes/methods we will instrument? It seems like the classname that is output should be an input. In production it's the same as the class we are reading, but in tests it can be something different.

Reply:
One `Instrumenter` handles all classes. It prefixes all of them.

Your way sounds sort of neat tho.